In [14]:
# Import all of the things you need to import!
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
count_vectorizer = CountVectorizer()

# Homework 14 (or so): TF-IDF text analysis and clustering

Hooray, we kind of figured out how text analysis works! Some of it is still magic, but at least the **TF** and **IDF** parts make a little sense. Kind of. Somewhat.

No, just kidding, we're *professionals* now.

## Investigating the Congressional Record

The [Congressional Record](https://en.wikipedia.org/wiki/Congressional_Record) is more or less what happened in Congress every single day. Speeches and all that. A good large source of text data, maybe?

Let's pretend it's totally secret but we just got it leaked to us in a data dump, and we need to check it out. It was leaked from [this page here](http://www.cs.cornell.edu/home/llee/data/convote.html).

In [1]:
# If you'd like to download it through the command line...
!curl -O http://www.cs.cornell.edu/home/llee/data/convote/convote_v1.1.tar.gz

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 9607k  100 9607k    0     0   226k      0  0:00:42  0:00:42 --:--:--  229k


In [2]:
# And then extract it through the command line...
!tar -zxf convote_v1.1.tar.gz

You can explore the files if you'd like, but we're going to get the ones from `convote_v1.1/data_stage_one/development_set/`. It's a bunch of text files.

In [3]:
# glob finds files matching a certain filename pattern
import glob

# Give me all the text files
paths = glob.glob('convote_v1.1/data_stage_one/development_set/*')
paths[:5]

['convote_v1.1/data_stage_one/development_set/052_400239_1479011_ROY.txt',
 'convote_v1.1/data_stage_one/development_set/493_400388_2346095_ROY.txt',
 'convote_v1.1/data_stage_one/development_set/199_400133_2013013_DOY.txt',
 'convote_v1.1/data_stage_one/development_set/493_400436_2346101_DON.txt',
 'convote_v1.1/data_stage_one/development_set/421_400338_2010025_ROY.txt']

In [4]:
len(paths)

702

So great, we have 702 of them. Now let's import them.

In [7]:
speeches = []
for path in paths:
    with open(path) as speech_file:
        speech = {
            'pathname': path,
            'filename': path.split('/')[-1],
            'content': speech_file.read()
        }
    speeches.append(speech)
speeches_df = pd.DataFrame(speeches)
speeches_df.head()

content  \
0  mr. chairman , i am happy to yield 2 minutes t...   
1  mr. chairman , i yield myself such time as i m...   
2  mr. speaker , i thank the gentleman from wisco...   
3  mr. chairman , i thank my friend , the gentlem...   
4  madam speaker , i recently hosted roundtables ...   

                     filename  \
0  052_400239_1479011_ROY.txt   
1  493_400388_2346095_ROY.txt   
2  199_400133_2013013_DOY.txt   
3  493_400436_2346101_DON.txt   
4  421_400338_2010025_ROY.txt   

                                            pathname  
0  convote_v1.1/data_stage_one/development_set/05...  
1  convote_v1.1/data_stage_one/development_set/49...  
2  convote_v1.1/data_stage_one/development_set/19...  
3  convote_v1.1/data_stage_one/development_set/49...  
4  convote_v1.1/data_stage_one/development_set/42...

In class we had the `texts` variable. For the homework can just do `speeches_df['content']` to get the same sort of list of stuff.

**Take a look at the contents of the first 5 speeches**

In [21]:
speeches_df['content'].head(5)

0    mr. chairman , i am happy to yield 2 minutes t...
1    mr. chairman , i yield myself such time as i m...
2    mr. speaker , i thank the gentleman from wisco...
3    mr. chairman , i thank my friend , the gentlem...
4    madam speaker , i recently hosted roundtables ...
Name: content, dtype: object

# Doing our analysis

Use the `sklearn` package and a plain boring `CountVectorizer` to get a list of all of the tokens used in the speeches. If it won't list them all, that's ok! Make a dataframe with those terms as columns.

**Be sure to include English-language stopwords**

In [25]:
count_vectorizer = CountVectorizer(stop_words='english')
X = count_vectorizer.fit_transform(speeches_df['content'])

In [26]:
X

<702x9106 sparse matrix of type '<class 'numpy.int64'>'
	with 56106 stored elements in Compressed Sparse Row format>

In [27]:
X.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       ..., 
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [28]:
pd.DataFrame(X.toarray())

0     1     2     3     4     5     6     7     8     9     ...   9096  \
0       0     0     0     0     0     0     0     0     0     0  ...      0   
1       0     0     0     0     0     0     0     0     0     0  ...      0   
2       1     0     0     0     0     0     0     0     0     0  ...      0   
3       0     0     0     0     0     0     0     0     0     0  ...      0   
4       0     0     0     0     0     0     0     0     0     0  ...      0   
5       0     0     0     0     0     0     0     0     0     0  ...      0   
6       0     0     0     0     0     0     0     0     0     0  ...      0   
7       0     0     0     0     0     0     0     0     0     0  ...      0   
8       3     0     0     0     0     0     0     0     0     0  ...      0   
9       1     0     0     0     0     0     0     0     0     0  ...      0   
10      1     0     0     0     0     0     1     0     0     0  ...      0   
11      0     0     0     0     0     0     0     0     0     0  ...      0   
12      0     0     0     0     0     0     0     0     0     0  ...      0   
13      0     0     0     0     0     0     0     0     0     0  ...      0   
14      0     0     0     0     0     0     0     0     0     0  ...      0   
15      0     0     0     0     0     0     0     0     0     0  ...      0   
16      0     0     0     0     0     0     0     0     0     0  ...      0   
17      2     0     0     0     0     0     0     0     0     0  ...      0   
18      0     0     0     0     0     0     0     0     0     0  ...      0   
19      0     0     0     0     0     0     0     0     0     0  ...      0   
20      0     0     0     0     0     0     0     0     0     0  ...      0   
21      0     0     0     0     0     0     0     0     0     0  ...      0   
22      2     0     0     0     0     0     0     0     0     0  ...      0   
23      0     0     0     0     0     0     0     0     0     0  ...      0   
24      0     0     0     0     0     0     0     0     0     0  ...      0   
25      0     0     0     0     0     0     0     0     0     0  ...      0   
26      0     0     0     0     0     0     0     0     0     0  ...      0   
27      0     0     0     0     0     0     0     0     0     0  ...      0   
28      1     0     0     0     0     0     0     0     0     0  ...      0   
29      0     0     0     0     0     0     0     0     0     0  ...      0   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...    ...   
672     0     0     0     0     0     0     0     0     0     0  ...      0   
673     0     0     0     0     0     0     0     0     0     0  ...      0   
674     0     0     0     0     0     0     0     0     0     0  ...      0   
675     0     0     0     0     0     0     0     0     0     0  ...      0   
676     0     0     0     0     0     0     0     0     0     0  ...      0   
677     0     0     0     0     0     0     0     0     0     0  ...      0   
678     0     0     0     0     0     0     0     0     0     0  ...      0   
679     0     0     0     0     0     0     0     0     0     0  ...      0   
680     0     0     0     0     0     0     0     0     0     0  ...      0   
681     0     0     0     0     0     0     0     0     0     0  ...      0   
682     0     0     0     0     0     0     0     0     0     0  ...      0   
683     0     0     0     0     0     0     0     0     0     0  ...      0   
684     0     0     0     0     0     0     0     0     0     0  ...      0   
685     0     0     0     0     0     0     0     0     0     0  ...      0   
686     0     0     0     0     0     0     0     0     0     0  ...      0   
687     0     0     0     0     0     0     0     0     0     0  ...      0   
688     0     0     0     0     0     0     0     0     0     0  ...      0   
689     0     0     0     0     0     0     0     0     0     0  ...      0   
690     0     0     0     0     0     0     0     0    

Okay, it's **far** too big to even look at. Let's try to get a list of features from a new `CountVectorizer` that only takes the top 100 words.

Now let's push all of that into a dataframe with nicely named columns.

Everyone seems to start their speeches with "mr chairman" - how many speeches are there total, and many don't mention "chairman" and how many mention neither "mr" nor "chairman"?

What is the index of the speech thank is the most thankful, a.k.a. includes the word 'thank' the most times?

If I'm searching for `China` and `trade`, what are the top 3 speeches to read according to the `CountVectoriser`?

Now what if I'm using a `TfidfVectorizer`?

**What's the content of the speeches?** Here's a way to get them:

In [ ]:
# index 0 is the first speech, which was the first one imported.
paths[0]

In [ ]:
# Pass that into 'cat' using { } which lets you put variables in shell commands
# that way you can pass the path to cat
!cat {paths[0]}

**Now search for something else!** Another two terms that might show up. `elections` and `chaos`? Whatever you thnik might be interesting.

# Enough of this garbage, let's cluster

Using a **simple counting vectorizer**, cluster the documents into **eight categories**, telling me what the top terms are per category.

Using a **term frequency vectorizer**, cluster the documents into **eight categories**, telling me what the top terms are per category.

Using a **term frequency inverse document frequency vectorizer**, cluster the documents into **eight categories**, telling me what the top terms are per category.

**Which one do you think works the best?**

# Harry Potter time

I have a scraped collection of Harry Potter fanfiction at https://github.com/ledeprogram/courses/raw/master/algorithms/data/hp.zip.

I want you to read them in, vectorize them and cluster them. Use this process to find out **the two types of Harry Potter fanfiction**. What is your hypothesis?